05_testing_and_ci/python_testing_demo.md

```diff
@@ -0,0 +1,58 @@
+# Notes for Demos of Python Testing Frameworks
+
+Example code is in [05_testing_and_ci/examples](https://github.com/Simulation-Software-Engineering/Lecture-Material/blob/main/05_testing_and_ci/examples)
+
+## Software code used
+
+- The file `operations.py` consists of two functions `find_max` and `find_average` which calculate the maximum and average of elements of a list. The `main()` routine in the file applies the functions to a list and prints the output.
+- `main()` function in `operations.py` has assertion statements to check if the correct data type is passed to specific functions.
+- Assertion statements are the most basic way of testing code and are also used in unit and integration testing as demonstrated here.
+- Tests are written in the file `test_operations.py`. The `test_*` prefix in the name is required so that pytest detects the file as a testing file. Suffix form `*_test.py` also works.
+- In all there are two unit tests, one integration test and one regression test.
+- The unit tests test the individual functions `find_max` and `find_average`.
+- The integration test triggers both the functions `find_max` and `find_average` and checks that the average is less than the maximum, something that should always be true for a set of numbers.
+- The regression test first reads an old data set and an average value from a CSV file. Then the function `find_average` is run with the old data set and the new average value is compared to the old one.
+
+## pytest
+
+- pytest is installed by: `pip install pytest`.
+- All tests can be run using the command-line tool called `pytest`.
+- **Comparing floating point variables** needs to be handled in functions like `find_average` and is done using `pytest.approx(value, abs)`.
+- Even if one test fails, pytest runs all the tests and gives a report on the failing test. The assertion failure report generated my pytest is also more detailed than the usual Python assertion report.
+- pytest can run unittest suites directly using `pytest tests`.
+- pytest is able to detect tests in several forms of folder structures, and the folder structures have advantages and disadvantages. More information on this is in the [documentation](https://docs.pytest.org/en/6.2.x/goodpractices.html#choosing-a-test-layout-import-rules).
+- Example:
+
+```
+setup.py
+my_software/
+    __init__.py
+    main.py
+    feature.py
+    test/
+        __init__.py
+        test_one.py
+        test_two.py
+        ...
+```
+
+## unittest
+
+- Base class `unittest.TestCase` is used to create a test suite consisting of all the tests of a software.
+- Each test is now a function of a class which is derived from the class `unittest.TestCase`.
+- The same tests are implemented using `unittest` in the file `test_operations_unittests.py` as functions of a class `TestOperations`.
```

Comment:
I would try to be more specific here to avoid confusion or misunderstandings.
```suggestion
- The same tests as for `pytest` are implemented using `unittest` in the file `test_operations_unittests.py`. The tests are functions of a class named `TestOperations` which tests our mathematical operations. The class `TestOperations` is derived from `unittest.TestCase`.
```